pulp_rpm/app/serializers.py

```diff
@@ -331,85 +343,189 @@ class Meta:
+            )
+            if 'shortname' not in collection:
+                collection['shortname'] = collection.pop("short", "")
+            coll = UpdateCollection(**collection)
+            coll.update_record = update_record
+            update_collections_to_save.append(coll)
+            for package in packages:
+                pkg = UpdateCollectionPackage(**package)
+                pkg.update_collection = coll
+                update_collection_packages_to_save.append(pkg)
+        for reference in references:
+            reference[PULP_UPDATE_REFERENCE_ATTRS.HREF] = reference.pop(
+                CR_UPDATE_REFERENCE_ATTRS.HREF
+            )
+            reference[PULP_UPDATE_REFERENCE_ATTRS.ID] = reference.pop(
+                CR_UPDATE_REFERENCE_ATTRS.ID
+            )
+            reference[PULP_UPDATE_REFERENCE_ATTRS.TITLE] = reference.pop(
+                CR_UPDATE_REFERENCE_ATTRS.TITLE
+            )
+            reference[PULP_UPDATE_REFERENCE_ATTRS.TYPE] = reference.pop(
+                CR_UPDATE_REFERENCE_ATTRS.TYPE
+            )
+            ref = UpdateReference(**reference)
+            ref.update_record = update_record
+            update_references_to_save.append(ref)
+
+        if update_collections_to_save:
+            UpdateCollection.objects.bulk_create(update_collections_to_save)
+        if update_collection_packages_to_save:
+            UpdateCollectionPackage.objects.bulk_create(update_collection_packages_to_save)
+        if update_references_to_save:
+            UpdateReference.objects.bulk_create(update_references_to_save)
+
+        return update_record
+
+    def validate(self, data):
+        """
+        Read a file for a JSON data and validate a UpdateRecord data.
+
+        Also change 'from' to 'fromstr' if exists as this is usually handle by
+        createrepo_c which is not used here.
+        """
+        update_record_data = dict()
+        if 'file' in data:
+            update_record_data.update(json.loads(data['file'].read()))
+            update_record_data.update(data)
+        elif 'artifact' in data:
+            update_record_data.update(json.loads(data['artifact'].file.read()))
+            update_record_data.update(data)
+        else:
+            raise serializers.ValidationError("Only creation with file or artifact is allowed.")
+
+        update_record_data[PULP_UPDATE_RECORD_ATTRS.FROMSTR] = update_record_data.pop(
+            'from', update_record_data.get(PULP_UPDATE_RECORD_ATTRS.FROMSTR, "")
+        )
+        update_record_data[PULP_UPDATE_RECORD_ATTRS.ISSUED_DATE] = update_record_data.pop(
+            'issued', update_record_data.get(PULP_UPDATE_RECORD_ATTRS.ISSUED_DATE, "")
+        )
+        update_record_data[PULP_UPDATE_RECORD_ATTRS.UPDATED_DATE] = update_record_data.pop(
+            'updated', update_record_data.get(PULP_UPDATE_RECORD_ATTRS.UPDATED_DATE, "")
+        )
+
+        validated_data = super().validate(update_record_data)
```

Comment:
i suggest moving this in the beginning so core validation occurs right away

Reply:
I do not agree as in this part few fields must be renamed (as in the JSON there are differences e.g. 'fromstr' in JSON is just 'from'.

Reply:
i see